### PR TITLE
[ECR-2319] frontend updated to support binary transactions

### PR DESCRIPTION
**Pull request by qvantor**

### CODE REVIEW

File: exonum-java-binding-cryptocurrency-demo/frontend/server.js [hunk 1]
```diff
@@ -1,44 +1,36 @@
...
-var explorerRoot = argv.explorerRoot;
-var apiRoot = argv.apiRoot;
+var argv = require('yargs-parser')(process.argv.slice(2))
+var port = argv.port
+var explorerRoot = argv.explorerRoot
+var apiRoot = argv.apiRoot
 
 if (typeof port === 'undefined') {
-  throw new Error('--port parameter is not set.');
+  throw new Error('--port parameter is not set.')
```
Comment (bullet-tooth) [inline exonum-java-binding-cryptocurrency-demo/frontend/server.js]: maybe we need to add `jslint` to the project to avoid such things?

Comment (KateNegrienko) [reply exonum-java-binding-cryptocurrency-demo/frontend/server.js]: Could you change var to let or const?

File: exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js [hunk 2]
```diff
@@ -2,26 +2,25 @@ import * as Exonum from 'exonum-client'
...
+let TransferTransactionProtobuf = new Type('TransferTransaction').add(new Field('seed', 1, 'int64'))
+TransferTransactionProtobuf.add(new Field('senderId', 2, 'bytes'))
+TransferTransactionProtobuf.add(new Field('recipientId', 3, 'bytes'))
+TransferTransactionProtobuf.add(new Field('amount', 4, 'int64'))
 
 // add types to protobuf root
-var root = new Root();
-root.define("CreateTransactionProtobuf").add(CreateTransactionProtobuf);
-root.define("TransferTransactionProtobuf").add(TransferTransactionProtobuf);
+var root = new Root()
```
Comment (KateNegrienko) [inline exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js]: var -> let or const?

File: exonum-java-binding-cryptocurrency-demo/frontend/server.js (hunk 1 above)
Comment (qvantor) [reply exonum-java-binding-cryptocurrency-demo/frontend/server.js]: @bullet-tooth we already started a process of moving all project to same linting rules

File: exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js [hunk 3]
```diff
@@ -87,103 +86,90 @@ function createHeader(messageId) {
...
         // serialize and append message body
-        const body = CreateTransactionProtobuf.encode(data).finish();
-
-        body.forEach(element => {
-          buffer.push(element)
-        });
+        const body = CreateTransactionProtobuf.encode(data).finish()
+        const bufferWithBody = appendBuffer(new Uint8Array(buffer), body)
 
         // calculate payload and insert it into buffer
```
Comment (dmitry-timofeev) [inline exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js]: "Update payload _size_ field in the header" (to be more specific)?

File: exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js [hunk 4]
```diff
@@ -87,103 +86,90 @@ function createHeader(messageId) {
...
-        body.forEach(element => {
-          buffer.push(element)
-        });
+        // serialize and append message body
+        const body = TransferTransactionProtobuf.encode(data).finish()
+        const bufferWithBody = appendBuffer(new Uint8Array(buffer), body)
 
         // calculate payload and insert it into buffer
-        Exonum.Uint32.serialize(buffer.length + SIGNATURE_LENGTH, buffer, PAYLOD_SIZE_OFFSET)
+        Exonum.Uint32.serialize(bufferWithBody.length + SIGNATURE_LENGTH, bufferWithBody, PAYLOD_SIZE_OFFSET)
```
Comment (dmitry-timofeev) [inline exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js]: I'd consider extracting in a function and using it in both methods
```
fun appendBodyToHeader(header: byte[], body: byte[]) {
  var unsignedMessage = header + body
  header[PAYLOAD_SIZE_OFFSET] = unsignedMessage.size()
}
```

File: exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js [hunk 5]
```diff
@@ -87,103 +86,90 @@ function createHeader(messageId) {
...
-        data.recipientId = receiver
-        return axios.post(TX_URL, {
-          protocol_version: PROTOCOL_VERSION,
-          service_id: SERVICE_ID,
-          message_id: TX_TRANSFER_ID,
-          signature: signature,
-          body: data
-        }).then(response => waitForAcceptance(keyPair.publicKey, response.data))
+        const signature = Exonum.sign(keyPair.secretKey, bufferWithBody)
+        const signBuffer = appendBuffer(bufferWithBody, Exonum.hexadecimalToUint8Array(signature))
```
Comment (dmitry-timofeev) [inline exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js]: … or even
```
fun buildMessage(header: byte[], body: byte[], keyPair: KeyPair) {
  var unsignedMessage = header + body
  header[PAYLOAD_SIZE_OFFSET] = unsignedMessage.size()
  var signature = sign(unsignedMessage)
  ret unsignedMessage + signature
}
```

File: exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js [hunk 6]
```diff
@@ -87,103 +86,90 @@ function createHeader(messageId) {
...
-
-        body.forEach(element => {
-          buffer.push(element)
-        });
+        const body = CreateTransactionProtobuf.encode(data).finish()
+        const bufferWithBody = appendBuffer(new Uint8Array(buffer), body)
 
         // calculate payload and insert it into buffer
-        Exonum.Uint32.serialize(buffer.length + SIGNATURE_LENGTH, buffer, PAYLOD_SIZE_OFFSET)
+        Exonum.Uint32.serialize(bufferWithBody.length + SIGNATURE_LENGTH, bufferWithBody, PAYLOD_SIZE_OFFSET)
```
Comment (dmitry-timofeev) [inline exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js]: PAYLOAD?

File: exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js [hunk 7]
```diff
@@ -87,103 +86,90 @@ function createHeader(messageId) {
...
+      generateSeed () {
         return bigInt.randBetween(0, MAX_VALUE).valueOf()
-      },  
-
-      createWallet(keyPair, balance) {
+      },
 
+      createWallet (keyPair, balance) {
         // serialize and append message header
         let buffer = createHeader(TX_WALLET_ID)
```
Comment (DENMROOT) [inline exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js]: Rename to header if it's header

File: exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js (hunk 5 above)
Comment (DENMROOT) [reply exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js]: +1 for naming like  unsignedMessage, message, signature

File: exonum-java-binding-cryptocurrency-demo/frontend/server.js (hunk 1 above)
Comment (qvantor) [reply exonum-java-binding-cryptocurrency-demo/frontend/server.js]: @bullet-tooth sure

**Comment by qvantor** [comment]
@DENMROOT @dmitry-timofeev @bullet-tooth @KateNegrienko Updated

File: exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js [hunk 8]
```diff
@@ -86,13 +86,26 @@ function createHeader (messageId) {
...
   tmp.set(new Uint8Array(buffer1), 0)
   tmp.set(new Uint8Array(buffer2), buffer1.byteLength)
   return tmp
 }
 
+const buildMessage = (header, body, keyPair) => {
+  const unsignedMessage = appendBuffer(new Uint8Array(header), body)
+
+  // calculate payload and insert it into buffer
+  Exonum.Uint32.serialize(unsignedMessage.length + SIGNATURE_LENGTH, unsignedMessage, PAYLOD_SIZE_OFFSET)
```
Comment (dmitry-timofeev) [inline exonum-java-binding-cryptocurrency-demo/frontend/src/plugins/blockchain.js]: Typo: `PAYLOAD_SIZE_OFFSET`

**Pull request merged**